docs/source/miscellaneous/doc/adr/0012-batch-timestamp-requests-on-the-client-side.md

```diff
@@ -0,0 +1,46 @@
+# 12. Batch timestamp requests on the client side
+
+Date: 26/09/2017
+
+## Status
+
+Accepted
+
+## Context
+
+A TimeLock cluster can serve multiple services. Under these circumstances, each TimeLock client node can separately
+open connections to TimeLock and request timestamps and locks. As timestamp and lock requests are typically very common
+(for instance, a single write transaction needs to contact TimeLock multiple times), TimeLock tends to experience
+relatively heavy loads. This affects overall throughput, as TimeLock takes longer to respond to requests.
+
+The `PersistentTimestampService` supports querying for a contiguous range of timestamps. This functionality has
+already been used in production, in Palantir's large internal product.
+
+## Decision
+
+- Use the `RequestBatchingTimestampService` to coalesce timestamp requests to TimeLock, and enable it by default.
+  This means that a TimeLock client will, at any time, have at most one request in flight to TimeLock.
```

Comment:
"...at most one **timelock** request in flight..."

Reply:
Actually, now you mention it, this should be "...at most one **timestamp** request in flight" - lock service requests aren't batched in any way.

Reply:
Actually, I got it wrong in the correction!  I meant **timestamp** request.  Lock requests are unaffected by this change.